### PR TITLE
fluent-plugin-loki: Restructuring and CI

### LINKED ISSUE #1724: CI steps to push fluentd image with Loki output plugin
**Is your feature request related to a problem? Please describe.**

Currently we have a docker image that embed fluentd and our Loki plugin. But we don't provide a built version that can be easily be pulled.

**Describe the solution you'd like**
We should build a new image on each merge to master, like we do for other docker images.

Bonus point it would be great to also automatically push to gem repository.

### PR DESCRIPTION
**Pull request by Skeen**
**What this PR does / why we need it**:
This PR:
1. Moves `fluentd/fluent-plugin-grafana-loki` to `cmd/fluentd`, as to follow the standard for the other integrations.
2. Integrates it with the `Makefile`, such that the commands `fluentd-plugin`, `fluentd-image`, `fluentd-push` and `fluentd-test` are added, as they are for the `fluent-bit` plugin.
3. Adds a drone-ci job to generate and keep the FluentD images up to date.

**Which issue(s) this PR fixes**:
Fixes #1724.

**Special notes for your reviewer**:
I have removed the dependency on systemd, as it seems to have only been needed for the specific configuration in `systemd.conf`, assuming a user needs this, they will need to create a derivative image with this as base image.

**Checklist**
- [ ] Documentation added
- [ ] Tests updated

### CODE REVIEW

**Comment by cyriltovena** [comment]
Wow we really needed that !! I’ll take a look this week, thank you !

**Comment by adityacs** [comment]
@Skeen Thanks for this PR.

@cyriltovena I think we should include publishing the fluentd gem to gem repository in this PR and also we can move both fluentbit and fluentd under `plugins` folder as discussed here #1989 WDYT?

**Comment by Skeen** [comment]
@adityacs as for publishing the Flutentd gem, I'm alfraid that I won't be able to help with that. I don't have any experience with Ruby.

Also, do we want to move them right away, or would we rather do it using two PRs:
1. This PR, which unifies the structure
2. A new PR, which restructures and moves the plugins to a seperate folder.

I'd personally argue for two seperate PRs, mainly to keep changes simple and traceable.

**Comment by cyriltovena** [comment]
It's ok for the gem, I can take over.

Yeah @adityacs this is a good idea, but let's send another PR for it.

**Comment by adityacs** [comment]
@cyriltovena Sure

File: .drone/drone.jsonnet [hunk 1]
```diff
@@ -84,6 +84,30 @@ local fluentbit() = pipeline('fluent-bit-amd64') + arch_image('amd64', 'latest,m
...
+      depends_on: ['image-tag'],
+      when: condition('exclude').tagMaster,
+      settings+: {
+        dry_run: true,
+        repo: 'grafana/fluent-plugin-loki',
+      },
+    },
+  ] + [
+    // publish for tag or master
+    docker('amd64', 'fluent-bit') {
```
Comment (cyriltovena) [inline .drone/drone.jsonnet:100]:
```suggestion
    docker('amd64', 'fluentd') {
```

File: Makefile [hunk 2]
```diff
@@ -522,6 +543,7 @@ ifeq ($(BUILD_IN_CONTAINER),true)
 		$(IMAGE_PREFIX)/loki-build-image:$(BUILD_IMAGE_VERSION) $@;
 else
 	drone jsonnet --stream --format -V __build-image-version=$(BUILD_IMAGE_VERSION) --source .drone/drone.jsonnet --target .drone/drone.yml
+	sed -i '1s/^/\# Generated by jsonnet from .drone\/drone.jsonnet\n/' .drone/drone.yml
```
Comment (cyriltovena) [inline Makefile:546]: Thanks ! 🥇

File: cmd/fluentd/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+RUN apt-get update && apt-get install -y --no-install-recommends libjemalloc2
+
+COPY cmd/fluentd/docker/Gemfile* /fluentd/
+COPY cmd/fluentd/docker/entrypoint.sh /fluentd/entrypoint.sh
+COPY cmd/fluentd/docker/conf/ /fluentd/etc/loki/
+
+ENV FLUENTD_CONF="/fluentd/etc/loki/fluentd.conf"
+ENV FLUENTD_OPT=""
+
+ENV LOKI_URL "https://logs-us-west1.grafana.net"
```
Comment (cyriltovena) [inline cmd/fluentd/Dockerfile:29]:
```suggestion
ENV LOKI_URL "https://logs-prod-us-central1.grafana.net"
```

I need to send a PR to change this everywhere.

File: .drone/drone.jsonnet (hunk 1 above)
Comment (Skeen) [reply .drone/drone.jsonnet]: Thanks, I must have missed that!

**Comment by Skeen** [comment]
> I'm wondering if the default conf we are providing in the container is a good one.

Providing default configuration files are a difficult task as usage patterns are so diverse.

However if it was purely up to me, i think I'd just go with the default that ships with FluentD, appended with the `@include loki.conf` line.
The argument for this approach would be, that we do not in fact take a stand with regards to the default configuration, but rather outsource this decision to upstream, and as such, if people are displased with the default configuration, it is not our issue, but rather an issue to take with upstream.

**Comment by Skeen** [comment]
> No idea why but on the second run of `make fluentd-image` I'm getting
> 
> ```
> bundle install --gemfile=cmd/fluentd/Gemfile --path=cmd/fluentd/vendor/bundle
> You must use Bundler 2 or greater with this lockfile.
> make: *** [Makefile:425: fluentd-plugin] Error 20
> The command '/bin/sh -c make BUILD_IN_CONTAINER=false fluentd-plugin' returned a non-zero code: 2
> make: *** [fluentd-image] Error 2
> ```
I'm afraid I have little to no experience with Ruby or it's toolchains.

**Comment by cyriltovena** [comment]
Could not find a lock file, I'll figure this out.

> However changing the default configuration file in general (which is already done in this PR, by eliminating the systemd dependency), has the issue of breaking backwards compatability for users running with the default configuration file. I don't know if this is a big issue, but it could be.

Nobody is really using that image since it's wasn't pushed first, I'm not worried that you go with the suggestion you have, I like it.

**Comment by Skeen** [comment]
@cyriltovena Am I right in understanding, that you do not expect me to make other changes before this is mergeable, or did I miss something?

File: .drone/drone.yml [hunk 4]
```diff
@@ -514,6 +515,62 @@ steps:
...
+  when:
+    ref:
+      exclude:
+      - refs/heads/master
+      - refs/heads/k??
+      - refs/tags/v*
+  depends_on:
+  - image-tag
+
+- name: publish-fluent-bit-image
```
Comment (cyriltovena) [inline .drone/drone.yml:554]: Do you need to regen the file ? Why is this there ?

File: cmd/fluentd/docker/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    build:
+      context: ../../..
+      dockerfile: cmd/fluentd/Dockerfile
+    image: fluentd:loki
+    volumes:
+    - ./fluentd.conf:/fluentd/etc/fluent.conf
+    environment:
+      - LOKI_URL
+  # Read /var/log/syslog and send it to fluentd
+  fluentbit:
```
Comment (cyriltovena) [inline cmd/fluentd/docker/docker-compose.yml:14]: is this just for testing ?

File: cmd/fluentd/docker/conf/loki.conf [hunk 6]
```diff
@@ -0,0 +1,9 @@
+<match loki.**>
+  @type loki
+  url "#{ENV['LOKI_URL']}"
+  extra_labels {"env":"dev"}
```
Comment (cyriltovena) [inline cmd/fluentd/docker/conf/loki.conf:4]: maybe job="fluentd" is a better default ?

**Comment by cyriltovena** [comment]
> @cyriltovena Am I right in assuming, that you do not expect me to make other changes before this is mergeable, or did I miss something?

No just have a look at the question I have.

File: cmd/fluentd/docker/docker-compose.yml (hunk 5 above)
Comment (Skeen) [reply cmd/fluentd/docker/docker-compose.yml:14]: Yes

File: .drone/drone.yml (hunk 4 above)
Comment (Skeen) [reply .drone/drone.yml]: I forgot to regen the file, I just did it in commit a1b54cb

File: cmd/fluentd/docker/conf/loki.conf (hunk 6 above)
Comment (Skeen) [reply cmd/fluentd/docker/conf/loki.conf]: Done in 244f196

**Comment by cyriltovena** [comment]
If you're up for a bit of documentation improvement that would greatly help again Thank you for this !

**Pull request merged**